crates/ra_hir_ty/src/_match.rs

```diff
@@ -573,15 +573,22 @@ pub(crate) fn is_useful(
     matrix: &Matrix,
     v: &PatStack,
 ) -> MatchCheckResult<Usefulness> {
-    // Handle the special case of enums with no variants. In that case, no match
-    // arm is useful.
+    // Handle two special cases:
+    // - enum with no variants
+    // - `!` type
+    // In those cases, no match arm is useful.
     if let Ty::Apply(ApplicationTy { ctor: TypeCtor::Adt(AdtId::EnumId(enum_id)), .. }) =
         cx.infer[cx.match_expr].strip_references()
     {
         if cx.db.enum_data(*enum_id).variants.is_empty() {
             return Ok(Usefulness::NotUseful);
         }
     }
+    if let Ty::Apply(ApplicationTy { ctor: TypeCtor::Never, .. }) =
+        cx.infer[cx.match_expr].strip_references()
```

Comment:
maybe a `match` instead of two `if let`s?

Reply:
I've changed to match. I can merge branches even further, but imho code will be too comples.